### PR TITLE
Embed

### PR DESCRIPTION
**Pull request by ringgaard**
* Training of word embeddings from Wikipedia document using word2vec.
* Training of fact and category embeddings from Wikidata using dual encoder model.

### CODE REVIEW

File: sling/nlp/embedding/word-embeddings.cc [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+    int epoch = 0;
+    std::vector<int> words;
+
+    // Set up model compute instances.
+    myelin::Instance l0(model->GetCell(flow_.layer0));
+    myelin::Instance l1(model->GetCell(flow_.layer1));
+    myelin::Instance l0b(model->GetCell(flow_.layer0b));
+
+    int *features = l0.Get<int>(model->GetParameter(flow_.fv));
+    int *fend = features + 2 * window_;
```
Comment (rahul1980) [inline sling/nlp/embedding/word-embeddings.cc]: flow_.in_features instead of 2 * window_ ?

File: sling/nlp/embedding/embedding-model.cc [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+  l1_l0 = tf.Instance(layer0);
+  auto *h = tf.Ref(l1_l0, hidden);
+
+  // Output.
+  bool single = out_features == 1;
+  auto *embed = single ? tf.Gather(W1, target) : tf.GatherAvg(W1, target);
+  auto *output = tf.Dot(embed, h, dims);
+
+  // Loss.
+  loss = tf.Name(tf.Sub(label, tf.Sigmoid(output)), "loss");
```
Comment (rahul1980) [inline sling/nlp/embedding/embedding-model.cc]: Actually, this is not the loss (or loss_grad) but the likelihood, since you are doing gradient ascent later on.

File: sling/nlp/embedding/fact-embeddings.cc [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+      optimizer_mu_.Unlock();
+
+      // Check if we are done.
+      if (EpochCompleted()) break;
+    }
+  }
+
+  // Evaluate model.
+  bool Evaluate(int64 epoch, myelin::Network *model) override {
+    // Evaluate model.
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/nlp/embedding/fact-embeddings.cc]: fixed

File: sling/nlp/embedding/word-embeddings.cc (hunk 1 above)
Comment (ringgaard) [reply sling/nlp/embedding/word-embeddings.cc]: fixed
